src/java/main/org/apache/zookeeper/common/X509Util.java

```diff
@@ -160,43 +214,120 @@ public static X509KeyManager createKeyManager(String keyStoreLocation, String ke
+                pbParams.setRevocationEnabled(true);
+                System.setProperty("com.sun.net.ssl.checkRevocation", "true");
+                System.setProperty("com.sun.security.enableCRLDP", "true");
+                if (ocspEnabled) {
+                    Security.setProperty("ocsp.enable", "true");
+                }
+
+            } else {
+                pbParams.setRevocationEnabled(false);
+            }
+
+            // Revocation checking is only supported with the PKIX algorithm
+            TrustManagerFactory tmf = TrustManagerFactory.getInstance("PKIX");
+            tmf.init(new CertPathTrustManagerParameters(pbParams));
+
+            for (final TrustManager tm : tmf.getTrustManagers()) {
+                if (tm instanceof X509ExtendedTrustManager) {
+                    return new ZKTrustManager((X509ExtendedTrustManager) tm, hostnameVerificationEnabled, shouldVerifyClientHostname);
                 }
             }
             throw new TrustManagerException("Couldn't find X509TrustManager");
-        } catch (Exception e) {
-            throw new TrustManagerException(e);
+        } catch (IOException|CertificateException|NoSuchAlgorithmException|InvalidAlgorithmParameterException|KeyStoreException
+                 trustManagerCreationException) {
+            throw new TrustManagerException(trustManagerCreationException);
         } finally {
             if (inputStream != null) {
                 try {
                     inputStream.close();
-                } catch (IOException e) {}
+                } catch (IOException ioException) {
+                    LOG.info("failed to close TrustStore input stream", ioException);
+                }
             }
         }
     }
-}
\ No newline at end of file
+
+    public SSLSocket createSSLSocket() throws X509Exception, IOException {
+        SSLSocket sslSocket = (SSLSocket) getDefaultSSLContext().getSocketFactory().createSocket();
+        configureSSLSocket(sslSocket);
+
+        return sslSocket;
+    }
+
+    public SSLSocket createSSLSocket(Socket socket) throws X509Exception, IOException {
+        SSLSocket sslSocket = (SSLSocket) getDefaultSSLContext().getSocketFactory().createSocket(socket, null, socket.getPort(), true);
+        configureSSLSocket(sslSocket);
+
+        return sslSocket;
+    }
+
+    private void configureSSLSocket(SSLSocket sslSocket) {
+        if (cipherSuites != null) {
+            SSLParameters sslParameters = sslSocket.getSSLParameters();
+            LOG.debug("Setup cipher suites for client socket: {}", Arrays.toString(cipherSuites));
+            sslParameters.setCipherSuites(cipherSuites);
+            sslSocket.setSSLParameters(sslParameters);
+        }
+    }
+
+
+    public SSLServerSocket createSSLServerSocket() throws X509Exception, IOException {
+        SSLServerSocket sslServerSocket = (SSLServerSocket) getDefaultSSLContext().getServerSocketFactory().createServerSocket();
```

Comment:
If we were to cache the socket factory, we might get some performance gain.

Reply:
I doubt that there is a lot to be gained, because we typically keep quorum connections open for a long time and don't create new sockets frequently. I think it only happens when the leader changes or an offline node comes back online.

Also, is the factory thread-safe? You can only cache it if so. If it's not, then every thread would need its own and at that point it may not be worth it.

Reply:
I got a little confused with ClientX509Util/QuorumX509Util classes. Looks like we don't use `ClientX509Util` to setup the pipeline for ZooKeeper client connections, it's essentially only used in the quorum communication for server and client side of server-server connections.

In which case I agree with caching the `SocketFactory` would not give noticeable performance gain. 

I'm also wondering whether it makes sense to move the SSLEngine setup logic in `InitSSL()` of `NettyServerCnxnFactory` to ClientX509Util.

Reply:
I wouldn't touch `NettyServerCnxnFactory` in this PR since quorum SSL and client SSL are basically independent of each other.